xds/internal/xdsclient/transport/lrs/lrs_stream.go

```diff
@@ -0,0 +1,326 @@
+	"google.golang.org/grpc/xds/internal/xdsclient/transport"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3endpointpb "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	v3lrspb "github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v3"
+)
+
+// Any per-RPC level logs which print complete request or response messages
+// should be gated at this verbosity level. Other per-RPC level logs which print
+// terse output should be at `INFO` and verbosity 2.
+const perRPCVerbosityLevel = 9
+
+// Stream provides all the functionality associated with an LRS (Load Reporting
+// Service) stream on the client-side.
+type Stream struct {
+	// The following fields are initialized from arguments passed to the
+	// constructor and are read-only afterwards, and hence can be accessed
+	// without a mutex.
+	transport transport.Interface     // Transport to use for LRS stream.
+	backoff   func(int) time.Duration // Backoff after stream failures.
+	nodeProto *v3corepb.Node          // Identifies the gRPC application.
+	doneCh    chan struct{}           // To notify exit of LRS goroutine.
+	logger    *igrpclog.PrefixLogger
+
+	// Guards access to the below fields.
+	mu           sync.Mutex
+	cancelStream context.CancelFunc // Cancel the stream. If nil, the stream is not active.
+	refCount     int                // Number of interested parties.
+	lrsStore     *load.Store        // Store returned to user for pushing loads.
+}
+
+// StreamOpts holds the options for creating an lrsStream.
+type StreamOpts struct {
+	Transport transport.Interface     // xDS transport to create the stream on.
+	Backoff   func(int) time.Duration // Backoff after stream failures.
+	NodeProto *v3corepb.Node          // Node proto to identify the gRPC application.
+	LogPrefix string                  // Prefix to be used for log messages.
+}
+
+// NewStream creates a new LRS Stream with the provided options.
+//
+// The actual streaming RPC call is initiated when the first call to ReportLoad
+// is made, and is terminated when the last call to ReportLoad is canceled.
+func NewStream(opts StreamOpts) *Stream {
+	lrs := &Stream{
+		transport: opts.Transport,
+		backoff:   opts.Backoff,
+		nodeProto: opts.NodeProto,
+		lrsStore:  load.NewStore(),
+	}
+
+	l := grpclog.Component("xds")
+	lrs.logger = igrpclog.NewPrefixLogger(l, opts.LogPrefix+fmt.Sprintf("[lrs-stream %p] ", lrs))
+	return lrs
+}
+
+// ReportLoad returns a load.Store that can be used to report load, and a
+// cleanup function that should be called when the load reporting is no longer
+// needed.
+//
+// The first call to ReportLoad will start the LRS streaming call. Subsequent
+// calls will increment the reference count and return the same load.Store. The
+// cleanup function will stop the LRS stream when the last reference is removed.
```

Comment:
will drop a reference, and will stop the LRS stream when the last reference is removed. Just to keep it symmetric with your comment about incrementing ref count. Also the first call also increases ref count too.

Reply:
Done.